### PR TITLE
Added split package info in email orders.

### PR DESCRIPTION
**Pull request by navidabdi**
### Ticket
https://app.asana.com/0/30156186242982/1203020088063708/f

### CODE REVIEW

File: src/WooCommerce.php [hunk 1]
```diff
@@ -1509,8 +1511,46 @@ private static function getRepricingOptionsDropdownValues(\WC_Product $product)
...
+
+      $shipping_methods_row = '';
+      $count = 1;
+
+      foreach($shipping_methods as $shipping_method) {
+        // Calculate => Shipping Price + Tax + Symbol
+        $shipping_price = floatval($shipping_method->get_total());
+        $shipping_price_tax = floatval($shipping_method->get_data()['total_tax']);
+        $shipping_price_total = $shipping_price + $shipping_price_tax;
+        $shipping_price_total_with_symbol = $shipping_price_total . $currency;
```
Comment (creative-andrew) [inline src/WooCommerce.php:1539]: Let's create a new method that takes care of this following the single responsibility principle.

Comment (creative-andrew) [inline src/WooCommerce.php:1539]: Let's create a new method that takes care of this following the single responsibility principle.

Comment (creative-andrew) [inline src/WooCommerce.php:1539]: Let's create a new method that takes care of this following the single responsibility principle.

Comment (creative-andrew) [inline src/WooCommerce.php:1539]:
```suggestion
        // Calculate => Shipping Price + Tax + Symbol
        $shipping_price = floatval($shipping_method->get_total());
        $shipping_price_tax = floatval($shipping_method->get_data()['total_tax']);
        $shipping_price_total = $shipping_price + $shipping_price_tax;
        $shipping_price_total_with_symbol = $shipping_price_total . $currency;
```

Let's create a new method that takes care of this following the single responsibility principle.

File: src/WooCommerce.php [hunk 2]
```diff
@@ -1509,8 +1511,56 @@ private static function getRepricingOptionsDropdownValues(\WC_Product $product)
...
+  /**
+   * Separates the shipping methods in new lines in emails.
+   *
+   * @param  array $total_rows
+   * @param \WC_Order $order
+   * @return array
+   */
+  public static function woocommerce_get_order_item_totals($total_rows, \WC_Order $order, $tax_display):array {
+    $shipping_methods = $order->get_shipping_methods();
+    $currency = ' ' . get_woocommerce_currency_symbol();
```
Comment (creative-andrew) [inline src/WooCommerce.php:1541]: Let's also move this into the function

File: src/WooCommerce.php [hunk 3]
```diff
@@ -1509,8 +1511,56 @@ private static function getRepricingOptionsDropdownValues(\WC_Product $product)
...
+
+    if(count($shipping_methods) > 1) {
+
+      $shipping_methods_row = '';
+      $count = 1;
+
+      foreach($shipping_methods as $shipping_method) {
+        $shipping_price_total = self::getTotalShippingPrice($shipping_method);
+        $shipping_price_total_with_symbol = $shipping_price_total . $currency;
```
Comment (creative-andrew) [inline src/WooCommerce.php:1550]: Let's move this into the function

File: src/WooCommerce.php [hunk 4]
```diff
@@ -1509,8 +1511,56 @@ private static function getRepricingOptionsDropdownValues(\WC_Product $product)
...
 
+  /**
+   * Calculates shipping price + tax
+   *
+   * @param  object $shipping_method
+   * @return float
+   */
+  public static function getTotalShippingPrice(object $shipping_method):float {
+    $shipping_price = floatval($shipping_method->get_total());
+    $shipping_price_tax = floatval($shipping_method->get_data()['total_tax']);
```
Comment (creative-andrew) [inline src/WooCommerce.php:1526]: There is a method to get the total tax:
https://woocommerce.github.io/code-reference/files/woocommerce-includes-class-wc-order-item-shipping.html#source-view.243

File: src/WooCommerce.php [hunk 5]
```diff
@@ -1509,8 +1511,56 @@ private static function getRepricingOptionsDropdownValues(\WC_Product $product)
...
     return $dropdown_values;
   }
 
+  /**
+   * Calculates shipping price + tax
+   *
+   * @param  object $shipping_method
+   * @return float
+   */
+  public static function getTotalShippingPrice(object $shipping_method):float {
```
Comment (creative-andrew) [inline src/WooCommerce.php:1524]:
```suggestion
  public static function getTotalShippingPrice(\WC_Order_Item_Shipping $shipping_method):string {
```

File: src/WooCommerce.php [hunk 6]
```diff
@@ -1509,8 +1511,56 @@ private static function getRepricingOptionsDropdownValues(\WC_Product $product)
...
+    $shipping_price = floatval($shipping_method->get_total());
```
Comment (creative-andrew) [inline src/WooCommerce.php:1525]: Some shipping methods don't have a price, they are free, so let's add an if statement and return early if that is the case.

Comment (creative-andrew) [reply src/WooCommerce.php:1525]:
```suggestion
    $shipping_price = floatval($shipping_method->get_total());
    if (!$shipping_price) {
     return "";
    }
```

File: src/WooCommerce.php [hunk 7]
```diff
@@ -1509,8 +1511,59 @@ private static function getRepricingOptionsDropdownValues(\WC_Product $product)
...
+      foreach($shipping_methods as $shipping_method) {
+        $shipping_price_total_with_symbol = self::getTotalShippingPrice($shipping_method);
+
+        $positionen = $shipping_method->get_meta("Positionen") ?? '';
+        $shipping_methods_row .= "($count) ";
+        $shipping_methods_row .= '<strong>' . $shipping_method->get_name() . '</strong>';
+        $shipping_methods_row .= $positionen ? ': ' . $positionen : '';
+        $shipping_methods_row .= $shipping_price_total_with_symbol;
+        $count++;
+
```
Comment (creative-andrew) [inline src/WooCommerce.php:1561]:
```suggestion
```

File: src/WooCommerce.php [hunk 8]
```diff
@@ -1509,8 +1511,59 @@ private static function getRepricingOptionsDropdownValues(\WC_Product $product)
...
     return $dropdown_values;
   }
 
+  /**
+   * Calculates shipping price + tax
+   *
+   * @param  object $shipping_method
+   * @return string
+   */
+  public static function getTotalShippingPrice(object $shipping_method):string {
```
Comment (creative-andrew) [inline src/WooCommerce.php:1524]:
```suggestion
  public static function getTotalShippingPrice():string {
```

**Review by creative-andrew — CHANGES_REQUESTED** [review]
Final minor changes.

File: src/WooCommerce.php (hunk 8 above)
Comment (navidabdi) [reply src/WooCommerce.php:1524]: Hey @creative-andrew, we need the $shipping_method argument in order to get the total shipping price for each element:

`$shipping_price_total_with_symbol = self::getTotalShippingPrice($shipping_method);`

Comment (creative-andrew) [reply src/WooCommerce.php:1524]: Sorry I meant it like this 

`public static function getTotalShippingPrice( $shipping_method):string {`

without the object in front.

Comment (creative-andrew) [reply src/WooCommerce.php:1524]: Sorry I meant it like this 

`public static function getTotalShippingPrice( $shipping_method):string {`

without the object in front.

**Review by sun — COMMENTED** [review]
Hey @navidabdi, this looks like a nice first change proposal!  As I had some time today and coincidentally stumbled into the notification about this PR, I reviewed it in more detail.  Please check my comments and let me know if you have questions.

On a larger note, we've recently stopped to put each and everything into too big grouped class files like "WooCommerce", but instead, use a more fine-grained approach of features instead.  Because as you can imagine for example a plugin like shop-standards has quite a lot of customizations for all kinds of areas of WooCommerce…  In this case, the feature that we are touching would be "WooCommerceShippingPackages".

File: src/WooCommerceShippingPackages.php [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+   * Separates the shipping methods in new lines in emails.
+   *
+   * @param array $total_rows
+   *   Total rows of woocommerce order items.
+   * @param \WC_Order $order
+   *   A Wc_Order object.
+   *
+   * @return array
+   *   The shipping details including shipping method name, shipping costs,
+   *   taxes and currency sign.
```
Comment (creative-andrew) [inline src/WooCommerceShippingPackages.php:58]:
```suggestion
   * @implements woocommerce_get_order_item_totals
```

Comment (creative-andrew) [reply src/WooCommerceShippingPackages.php:58]: As mentioned here: https://github.com/netzstrategen/wordpress-shop-standards/pull/218#discussion_r1005524253

if you find a more suitable initial description for the method, you can also change it. IMO, it explains quite good what we are trying to accomplish.

File: src/WooCommerceShippingPackages.php [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+/**
+ * Shipping related settings and actions.
+ */
+class WooCommerceShippingPackages {
+
+  /**
+   * Initialize the needed functions and hooks.
+   */
+  public static function init() {
+    // Show multiple shipping methods separately in order emails.
```
Comment (creative-andrew) [inline src/WooCommerceShippingPackages.php:19]:
```suggestion
```
Let's remove this comment.

File: src/WooCommerceShippingPackages.php [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+  public static function woocommerce_get_order_item_totals($total_rows, \WC_Order $order): array {
+    $shipping_methods = $order->get_shipping_methods();
+
+    if (count($shipping_methods) > 1) {
+
+      $shipping_methods_row = '';
+      $count = 1;
+
+      foreach ($shipping_methods as $shipping_method) {
+        $shipping_price_total_with_symbol = self::getShippingCostsIncludingTaxes($shipping_method);
```
Comment (creative-andrew) [inline src/WooCommerceShippingPackages.php:62]:
```suggestion
        $shipping_price_total_with_symbol = self::getShippingCostsIncludingTaxes($shipping_method, $order);
```

File: src/WooCommerceShippingPackages.php [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+   * Calculates the shipping costs including tax.
+   *
+   * @param \WC_Order_Item_Shipping $shipping_method
+   *   A shipping item object derived from a WC_Order.
+   *
+   * @return string
+   *   The shipping costs including taxes, currency sign, and HTML line-breaks,
+   *   for output in the HTML order emails.
+   */
+  public static function getShippingCostsIncludingTaxes(\WC_Order_Item_Shipping $shipping_method): string {
```
Comment (creative-andrew) [inline src/WooCommerceShippingPackages.php:33]:
```suggestion
  public static function getShippingCostsIncludingTaxes(\WC_Order_Item_Shipping $shipping_method, \WC_Order $order): string {
```

File: src/WooCommerceShippingPackages.php [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+    if ($shipping_price) {
+      $currency = ' ' . get_woocommerce_currency_symbol();
+      $shipping_price = floatval($shipping_method->get_total());
+      $shipping_price_tax = floatval($shipping_method->get_total_tax());
+      $shipping_price_total = $shipping_price + $shipping_price_tax;
+      $shipping_price_total_with_symbol = ' - ' . $shipping_price_total . $currency;
```
Comment (creative-andrew) [inline src/WooCommerceShippingPackages.php:40]:
```suggestion
wc_price( $shipping_method->get_total() + $shipping_method->get_total_tax(), ['currency' => $order->get_currency()] );
```

File: src/WooCommerceShippingPackages.php [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+   * @param \WC_Order_Item_Shipping $shipping_method
+   *   A shipping item object derived from a WC_Order.
+   *
+   * @return string
+   *   The shipping costs including taxes, currency sign, and HTML line-breaks,
+   *   for output in the HTML order emails.
+   */
+  public static function getShippingCostsIncludingTaxes(\WC_Order_Item_Shipping $shipping_method): string {
+
+    if ($shipping_price) {
```
Comment (creative-andrew) [inline src/WooCommerceShippingPackages.php:35]:
```suggestion
    if ($shipping_price) {
```

I am not sure what are you checking here. In any case, you should check $shipping_method & $order.

File: src/WooCommerceShippingPackages.php (hunk 13 above)
Comment (creative-andrew) [reply src/WooCommerceShippingPackages.php:40]: I am just following how they are doing it here: https://woocommerce.github.io/code-reference/files/woocommerce-includes-abstracts-abstract-wc-order.html#source-view.2012

**Review by creative-andrew — APPROVED** [review]

**Pull request merged**